src/cart/src/cart/crt_context.c

```diff
@@ -1293,37 +1267,83 @@ crt_progress(crt_context_t crt_ctx, int64_t timeout,
 			hg_timeout = timeout;
 	}
 
-	while (true) {
+	/** loop until callback returns non-null value */
+	while ((rc = cond_cb(arg)) == 0) {
+		int ret;
+
+		rc = crt_hg_progress(&ctx->cc_hg_ctx, hg_timeout);
```

Comment:
we probably want to have a  comment here why we are calling crt_hg_progress() twice.
Also 'rc' here gets lost, do we want to process some fatal event here and break/return rc?

Reply:
Agreed, let's at least print an error if it fails.